src/main/perl/RuleBasedEditor.pm

```diff
@@ -0,0 +1,1154 @@
+Readonly my $RULE_CONDITION_ALWAYS          => 'ALWAYS';
+Readonly my $RULE_OPTION_SET_GLOBAL         => 'GLOBAL';
+
+
+# Export constants used to build rules
+# Needs to be updated when a constant is added or removed
+Readonly my @RULE_CONSTANTS => qw(
+  LINE_FORMAT_SH_VAR
+  LINE_FORMAT_ENV_VAR
+  LINE_FORMAT_KEY_VAL
+  LINE_FORMAT_KEY_VAL_SETENV
+  LINE_FORMAT_KEY_VAL_SET
+  LINE_VALUE_AS_IS
+  LINE_VALUE_BOOLEAN
+  LINE_VALUE_INSTANCE_PARAMS
+  LINE_VALUE_ARRAY
+  LINE_VALUE_HASH_KEYS
+  LINE_VALUE_STRING_HASH
+  LINE_VALUE_OPT_SINGLE
+  LINE_VALUE_OPT_UNIQUE
+  LINE_VALUE_OPT_SORTED
+  );
+
+
+our @EXPORT_OK;
+our %EXPORT_TAGS;
+push @EXPORT_OK, @RULE_CONSTANTS;
+$EXPORT_TAGS{rule_constants} = \@RULE_CONSTANTS;
+
+
+# Backup file extension
+Readonly my $BACKUP_FILE_EXT => ".old";
+
+=pod
+
+=head2 Public methods
+
+=over
+
+=item updateFile
+
+Update configuration file contents,  applying configuration rules.
+
+Arguments :
+    config_rules: config rules corresponding to the file to build
+    config_options: configuration parameters used to build actual configuration
+    options: a hash setting options to modify the behaviour of this function
+
+Supported entries for options hash:
+    always_rules_only: if true, apply only rules with ALWAYS condition (D: false)
+    remove_if_undef: if true, remove matching configuration line is rule condition is not met (D: false)
+
+Return value
+    sucess: 1
+    argument error: undef
+    
+=cut
+
+sub updateFile
+{
+    my $function_name = "updateConfigFile";
+    my ($self, $config_rules, $config_options, $parser_options) = @_;
+
+    unless ($config_rules) {
+        *$self->{LOG}->error("$function_name: 'config_rules' argument missing (internal error)");
+        return;
+    }
+    unless ($config_options) {
+        *$self->{LOG}->error("$function_name: 'config_options' argument missing (internal error)");
+        return;
+    }
+    unless (defined($parser_options)) {
+        *$self->{LOG}->debug(2, "$function_name: 'parser_options' undefined");
+        $parser_options = {};
+    }
+
+    $self->seek_begin();
+
+    # Check that config file has an appropriate header
+    Readonly my $INTRO_PATTERN => "# This file is managed by Quattor";
+    my $intro = "# This file is managed by Quattor - DO NOT EDIT lines generated by Quattor";
+    $self->add_or_replace_lines(
+                                qr/^$INTRO_PATTERN/,
+                                qr/^$intro$/,
+                                $intro . "\n#\n",
+                                BEGINNING_OF_FILE,
+                               );
+
+    $self->_apply_rules(
+                        $config_rules,
+                        $config_options,
+                        $parser_options
+                       );
+
+    return 1;
+}
+
+
+=pod
+
+=back
+
+=head2 Private methods
+
+=over
+
+=item formatAttrValue
+
+This function formats an attribute value based on the value format specified.
```

Comment:
die() is forbidden in a component... This is up to the caller to do what is necessary...